### PR TITLE
Refactor the AuthInterceptor

### PR DESCRIPTION
**Pull request by Nexcius**

### CODE REVIEW

File: core/src/main/java/com/schibsted/account/network/AuthInterceptor.kt [hunk 1]
```diff
@@ -81,40 +61,43 @@ class AuthInterceptor internal constructor(
...
-            if (result is AuthCheck.AuthCheckResult.Failed) {
-                Logger.error(TAG, { "Cannot perform authenticated request: ${result.reason}" })
-                throw AuthException("Cannot perform authenticated request: ${result.reason}")
-            }
+
+        // Preform security checks
+        val securityCheckResult = authSecurityCheck.securityCheck(originalRequest)
+        if ( securityCheckResult is AuthCheckResult.Failed ) {
+            Logger.error(TAG, { "Cannot perform authenticated request (ReqId:$reqId): ${securityCheckResult.reason}" })
+            throw AuthException("Cannot perform authenticated request (ReqId:$reqId): ${securityCheckResult.reason}")
```
Comment (Astach) [inline core/src/main/java/com/schibsted/account/network/AuthInterceptor.kt]: do we really need to log it if we already throw the exception?

File: core/src/main/java/com/schibsted/account/network/AuthInterceptor.kt [hunk 2]
```diff
@@ -81,40 +61,43 @@ class AuthInterceptor internal constructor(
...
         }
         Logger.verbose(TAG, { "Security checks passed for request (ReqId:$reqId)" })
 
         val token = user.token ?: throw AuthException("Cannot perform authenticated request (ReqId:$reqId) when the user is logged out")
+
+        // Add auth header if the URL is whitelisted
         val request = with(originalRequest.newBuilder()) {
             if (urlInWhitelist(originalUrl)) {
```
Comment (Astach) [inline core/src/main/java/com/schibsted/account/network/AuthInterceptor.kt]: we could replace urlInWhitelist method by
`private fun HttpUrl.isWhiteListed() = urlWhitelist.find { this.toString().startsWith(it) } != null` and write `originalUrl.isWhiteListed` is more readable

File: core/src/main/java/com/schibsted/account/network/AuthResponseCheck.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                override fun responseCheck(originalRequest: Request, response: Response): AuthCheckResult {
+                    val originalUrl = originalRequest.url()
+                    val responseUrl = response.request().url()
+
+                    if (whitelist.find { responseUrl.toString().startsWith(it) } == null) {
+                        return AuthCheckResult.Failed("URL not whitelisted")
+                    }
+
+                    if (responseUrl != originalUrl) {
+                        return AuthCheckResult.Failed("Current URL(${response.request().url().toString().safeUrl()}) does not match the original ${originalUrl.toString().safeUrl()}")
```
Comment (Astach) [inline core/src/main/java/com/schibsted/account/network/AuthResponseCheck.kt]: replace `response.request().url()` by `responseUrl`

File: core/src/main/java/com/schibsted/account/network/AuthResponseCheck.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    companion object {
+
+        fun default(whitelist: List<String>): AuthResponseCheck {
+            return object : AuthResponseCheck {
+                override fun responseCheck(originalRequest: Request, response: Response): AuthCheckResult {
+                    val originalUrl = originalRequest.url()
+                    val responseUrl = response.request().url()
+
+                    if (whitelist.find { responseUrl.toString().startsWith(it) } == null) {
```
Comment (Astach) [inline core/src/main/java/com/schibsted/account/network/AuthResponseCheck.kt]: could use the `url.isWhiteListed` I'm talking about

**Comment by zamzterz** [comment]
Closing due to inactivity.

**Pull request closed without merging**